python-checks/src/main/java/org/sonar/python/checks/StringFormat.java

```diff
@@ -0,0 +1,185 @@
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+
+public class StringFormat {
+
+  public static class ReplacementField {
+    private Consumer<Expression> validator;
+    private String mappingKey;
+
+    public ReplacementField(Consumer<Expression> validator, @Nullable String mappingKey) {
+      this.validator = validator;
+      this.mappingKey = mappingKey;
+    }
+
+    public void validateArgument(Expression expression) {
+      this.validator.accept(expression);
+    }
+
+    @CheckForNull
+    public String mappingKey() {
+      return mappingKey;
+    }
+  }
+
+  private List<ReplacementField> replacementFields;
+
+  private StringFormat(List<ReplacementField> replacementFields) {
+    this.replacementFields = replacementFields;
+  }
+
+  public List<ReplacementField> replacementFields() {
+    return this.replacementFields;
+  }
+
+  public int numExpectedArguments() {
+    return this.replacementFields.size();
+  }
+
+  public boolean hasPositionalFields() {
+    return this.replacementFields.stream().anyMatch(field -> field.mappingKey() == null);
+  }
+
+  public boolean hasNamedFields() {
+    return this.replacementFields.stream().anyMatch(field -> field.mappingKey() != null);
+  }
+
+  // See https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting
+  private static final Pattern PRINTF_PARAMETER_PATTERN = Pattern.compile(
+    "%" + "(?:\\((.*?)\\))?" + "([#\\-+0 ]*)?" + "([0-9]*|\\*)?" + "(?:\\.([0-9]*|\\*))?" + "([lLH])?" + "([A-Za-z]|%)");
+
+  private static final String PRINTF_NUMBER_CONVERTERS = "diueEfFgG";
+  private static final String PRINTF_INTEGER_CONVERTERS = "oxX";
+  private static final String PRINTF_VALID_CONVERTERS = PRINTF_NUMBER_CONVERTERS + PRINTF_INTEGER_CONVERTERS + "rsac";
+
+  public static Optional<StringFormat> createFromPrintfStyle(SubscriptionContext ctx, Tree tree, String input) {
+    List<ReplacementField> result = new ArrayList<>();
+    Matcher matcher = PRINTF_PARAMETER_PATTERN.matcher(input);
+
+    while (matcher.find()) {
+      String mapKey = matcher.group(1);
+      String conversionType = matcher.group(6);
+
+      if (conversionType.equals("%")) {
+        // If the conversion type is '%', we are dealing with a '%%'
+        continue;
+      }
+
+      char conversionTypeChar = conversionType.charAt(0);
+      if (PRINTF_VALID_CONVERTERS.indexOf(conversionTypeChar) == -1) {
+        ctx.addIssue(tree, String.format("Fix this formatted string's syntax; %%%c is not a valid conversion type.", conversionTypeChar));
+        return Optional.empty();
+      }
+
+      String width = matcher.group(3);
+      String precision = matcher.group(4);
+      if ("*".equals(width)) {
+        result.add(new ReplacementField(printfWidthOrPrecValidator(ctx), mapKey));
+      }
+      if ("*".equals(precision)) {
+        result.add(new ReplacementField(printfWidthOrPrecValidator(ctx), mapKey));
```

Comment:
As I understand, `*` must be used with sequential values only. Therefore I assume it would make sense to not pass the `mapKey` here. 

The current behavior is a bit misleading:
```
print("hello %(name)*s" % {"name": "John"}) # Noncompliant {{Replace this value with an integer as "*" requires.}}
#                                  ^^^^^^
print("hello %(name)*s" % {"name": 9}) # OK (FN)
```